google/cloud/forseti/services/inventory/storage.py

```diff
@@ -937,217 +906,50 @@ def write(self, resource):
 
         Args:
             resource (object): Resource object to store in db.
-
-        Raises:
-            Exception: If storage was opened readonly.
         """
-        if self.readonly:
-            raise Exception('Opened storage readonly')
-
-        resource_data = resource.data()
-        # Group members do not need to be checked if it already exists
-        # in Inventory, as a user can be members in multiple groups.
-        if resource_data.get('kind') != 'admin#directory#member':
-            previous_id = self._get_resource_id(resource)
-            if previous_id:
-                resource.set_inventory_key(previous_id)
-                self.update(resource)
+        # Use a lock to quickly check if this is a duplicate resource before
+        # updating the cache and proceeding.
+        with self._storage_lock:
+            if resource.get_full_resource_name() in self._wrote_resources:
+                LOGGER.warning('Duplicate Resource in inventory, skipping %s',
```

Comment:
Hmmm, are group members handled properly here? i.e. can group members in multiple groups can be added?

Reply:
Yes, groups members full names are like this (they are children of the group they are in): 
organization/111222333/gsuite_group/1013/gsuite_group_member/1025/
organization/111222333/gsuite_group/1014/gsuite_group_member/1025/

This is explicitly tested for in storage_test.py test_whether_resource_should_be_inserted_or_updated.